File: lib/note.rb
class Note
  
  attr_accessor :name, :prev_interval, :next_interval

  def initialize(name, prev_interval = nil, next_interval = nil)
    @name = name
    classify
  end

  def next_note
    Note.new Note.reference(self.name[0])[:next_note]
  end

  def prev_note
    name = self.name
    Note.new( name[1] ? name[0] : Note.reference(name)[:prev_note])
  end

  def next_half_step
    return self.next_note if self.next_interval == 0.5 
    if self.name[1]
      base = Note.reference(self.name[0])
      Note.new(base[:next_note])
    else
      Note.new(self.name + "#")
    end
  end

  def prev_half_step
    return self.prev_note if self.prev_interval == 0.5
    if self.name[1]
      Note.new(self.name[0])
    else
      Note.new( Note.reference(self.name)[:prev_note] + "#")
    end
  end

  def next_whole_step
    self.next_interval == 1 ? self.next_note : self.next_note.next_half_step
  end

  def prev_whole_step
    self.prev_interval == 1 ? self.prev_note : self.prev_note.prev_half_step
  end

  def classify
    if self.name[1]
      self.next_interval = self.prev_interval = 0.5
    else
      ref = Note.reference self.name
      self.next_interval, self.prev_interval = ref[:next_interval], ref[:prev_interval]
    end
  end

  def self.reference(name)
    case name
    when "A"
      {next_interval: 1, prev_interval: 1, next_note: "B", prev_note: "G"}
    when "B"
      {next_interval: 0.5, prev_interval: 1, next_note: "C", prev_note: "A"}
    when "C"
      {next_interval: 1, prev_interval: 0.5, next_note: "D", prev_note: "B"}
    when "D"
      {next_interval: 1, prev_interval: 1, next_note: "E", prev_note: "C"}
    when "E"
      {next_interval: 0.5, prev_interval: 1, next_note: "F", prev_note: "D"}
    when "F"
      {next_interval: 1, prev_interval: 0.5, next_note: "G", prev_note: "E"}
    when "G"
      {next_interval: 1, prev_interval: 1, next_note: "A", prev_note: "F"}
    end
  end

end

puts Note.new("A").next_half_step.name

File: java/Note.java
import java.util.*;
import java.lang.*;

public class Note {
  protected final String noteName;
  protected Double nextInterval;
  protected Note nextNote;
  protected Double prevInterval;
  protected Note prevNote;

  public Note(String noteName){
    this.noteName = noteName;


    // Determine intervals based on sharps
    if (this.noteName.length() > 1){
      this.nextInterval = 0.5;
      this.prevInterval = 0.5;
    } else {
      // Get data from lookUp static class
      String[] intervals = Note.lookUp(this.noteName);
      this.nextInterval = Double.parseDouble(intervals[0]);
      this.prevInterval = Double.parseDouble(intervals[1]);
    }
  }

  public static Map<String,String[]> fillMap(){

    // HashMap to keep track of each notes intervals and neighboring notes
    Map<String, String[]> intervals = new HashMap<String, String[]>();

    // Array of all possible notes
    String[] allNotes = {"A","A#","B","C","C#","D","D#","E","F","G","G#"};

    // Loop through every note and populate our hash map
    for (String i : allNotes ){
      intervals.put(i, new String[4]);
    };

    // Update the map so all notes are covered
    intervals.put("A", new String[] {"1","1","B","G"});
    intervals.put("B", new String[] {"0.5","1","C","A"});
    intervals.put("C", new String[] {"1","0.5","D","B"});
    intervals.put("D", new String[] {"1","1","E","C"});
    intervals.put("E", new String[] {"0.5","1","F","D"});
    intervals.put("F", new String[] {"1","0.5","G","E"});
    intervals.put("G", new String[] {"1","1","A","F"});

    return intervals;
  }

  public static String[] lookUp(String key){
    Map<String, String[]> intervals = fillMap();
    return intervals.get(key);
  }

  public Note nextNote(){
    String startingLetter = Character.toString(this.noteName.charAt(0));
    return new Note(Note.lookUp(startingLetter)[2]);
  }

  public Note prevNote(){
    String startingLetter = Character.toString(this.noteName.charAt(0));
    return new Note(Note.lookUp(startingLetter)[3]);
  }

  public Note nextHalfStep(){
    // If the next note a half step away, return it
    if (this.nextInterval == 0.5){ return this.nextNote(); };

    // Otherwise, if the length of the note name is > 1, we have a sharp
    if (this.noteName.length() > 1){
      String startingLetter = Character.toString(this.noteName.charAt(0));
      return new Note(Note.lookUp(startingLetter)[2]);
    } else {
      return new Note(this.noteName + "#");
    }
  }

  public Note nextWholeStep(){
    return (this.nextInterval == 1) ? nextNote() : nextNote().nextHalfStep();
  }

  public Note prevHalfStep(){
    // If the next note a half step away, return it
    if (this.nextInterval == 0.5){ return this.nextNote(); };

    // Otherwise, if the length of the note name is > 1, we have a sharp
    if (this.noteName.length() > 1){
      String startingLetter = Character.toString(this.noteName.charAt(0));
      return new Note(startingLetter);
    } else {
      return new Note(Note.lookUp(this.noteName)[3] + "#");
    }
  }

  public Note prevWholeStep(){
    return (this.nextInterval == 1) ? this.prevNote() : this.prevNote().prevHalfStep();
  }

  public String getNoteName(){
    return noteName;
  }

  public static void main(String[] args){

    Note aSharp = new Note("A#");
    System.out.println(aSharp.getNoteName());
    System.out.println(aSharp.nextNote().getNoteName());
    System.out.println(aSharp.prevNote().getNoteName());
    System.out.println(aSharp.nextWholeStep().getNoteName());
    System.out.println(aSharp.nextHalfStep().getNoteName());
    System.out.println(aSharp.prevWholeStep().getNoteName());
    System.out.println(aSharp.prevHalfStep().getNoteName());
  }

}